Rewrite/redesign data location checking

sifmelcara opened the pull request:
WIP. Need to fix tests and add more test cases.
Fixes #4382 , and #4319 can be fixed easily after this PR is merged.
Besides refactoring and fixing #4382, I also tried to enforce data location keyword whenever there is ambiguity. For example, `contract C { function f(bytes b) internal {} }` will cause type error now:
```
a.sol:1:25: Error: Location has to be memory or storage for reference type parameter. Use an explicit data location keyword to fix this error.
contract C { function f(bytes b) internal {} }
                        ^-----^
```
Not sure if this is a good idea though.

I will fix tests and add more test cases if everything looks ok.

chriseth commented:
Please note that in 2-3 weeks, it might be much easier to implement this enforcement, because we will disallow `var` and we will force users to always specify the data location.

sifmelcara commented:
Ok I will wait for the patch which disallow `var`.

chriseth commented:
Thanks, that is very much appreciated!

axic commented:
@sifmelcara removal of `var` was merged if you are interested in working on this :wink:

sifmelcara committed:
Refactor and enforce data location checking

Require user to specify data location when there are multiple
possible data locations for the variable declaration.

sifmelcara committed:
Add test case for data locations in function pointer

sifmelcara commented:
@axic I think this is ready to be reviewed.

sifmelcara commented:
Maybe I also need update codes in the documents? Will do that if changes made in this commit are considered ok.

axic commented:
Also #4527 fixed #4319. Since you have mentioned it would be easy to fix after this PR, is that still the case? Would you suggest a different fix than #4527?

axic commented:
Also it looks like a nice refactoring, please bear with us to review it. We have been a bit overwhelmed by the number of PRs.

sifmelcara commented:
@axic This PR also fixed the mapping data location enforcement (the fix was added after #4527 got merged). I fixes #4319 in the same way #4527 does, by reporting an error if the storage keyword is not given to local mapping variables.

axic commented:
It seems test are failing, were all the test expectations updated?

sifmelcara commented:
The test was failing because I didn't run cmdTests locally and did not fix sources in `docs/solidity-by-example.rst`. Now it should be fine.

chriseth commented:
I think there is no easy way to put it, but it seems there is a big overlap between this PR and #4518 - both started out fixing different issues in the data location area and then of course fixed the whole thing.

This PR nicely fixes the data location of function type variables, while I like how the other PR extracts the data location inference into its own function.

How shall we proceed? I fear the best way to solve this would be for e.g. me to create a new PR that picks elements from both your PRs (and of course also assigns authorship to both of you). What do you think?

I'm sorry that this happened, we should have been more careful.

chriseth commented:
Replaced by https://github.com/ethereum/solidity/pull/4738
